vllm/model_executor/layers/fused_moe/modular_kernel.py

```diff
@@ -1131,6 +1153,17 @@ def _finalize(
 
             receiver()
 
+        # ensure that any work enqueued on the shared_experts_stream is
+        # completed before the shared_output tensor is consumed
+        if (
+            self.shared_experts is not None
+            and use_shared_experts_stream
+            and self.shared_experts_stream is not None
+            and hidden_states.is_cuda
+            and torch.cuda.is_available()
+        ):
+            torch.cuda.current_stream().wait_stream(self.shared_experts_stream)
+
```

Comment:
Would it be a problem if the previous code errored out and the stream didn't get waited on?  Should this go in a `finally` block or use a context?

Reply:
You’re right that using a finally or context could make the error path more robust, but in this case if any of the preceding calls fail we also never consume `shared_output` or return from `_finalize`, so skipping the wait_stream on an exception doesn’t introduce a correctness issue. Given that and to keep this change minimal/consistent with the existing overlap pattern, I’d prefer to keep the current structure for now and can revisit a try/finally refactor in a follow‑up if we see a concrete case where it helps.

Reply:
Can you put this in a helper function too?  Then all the shared experts handling will be isolated.